boot/bootstate20.go

```diff
@@ -356,31 +356,96 @@ func (envbks *envRefExtractedKernelBootloaderKernelState) setNextKernel(sn snap.
 }
 
 //
-// kernel snap methods
+// bootStateUpdate for 20 methods
 //
 
-// bootState20Kernel implements the bootState and bootStateUpdate interfaces for
-// kernel snaps on UC20. It is used for setNext() and markSuccessful() - though
-// note that for markSuccessful() a different bootStateUpdate implementation is
-// returned, see bootState20MarkSuccessful
-type bootState20Kernel struct {
-	bks bootloaderKernelState20
+type bootCommitTask func() error
+
+// bootState20Kernel implements the bootStateUpdate interface for both kernel
+// and base snaps on UC20.
+type bootStateUpdate20 struct {
+	// tasks to run before the modeenv has been written
+	preModeenvTasks []bootCommitTask
+
+	// the modeenv that was read from the state
+	modeenv *Modeenv
 
-	// the kernel snap that was booted for markSuccessful()
-	bootedKernelSnap snap.PlaceInfo
+	// the modeenv that will be written out in commit
+	writeModeenv *Modeenv
 
-	// the kernel snap to try for setNext()
-	nextKernelSnap snap.PlaceInfo
+	// tasks to run after the modeenv has been written
+	postModeenvTasks []bootCommitTask
+}
 
-	// the kernel_status to commit during commit()
-	commitKernelStatus string
+func newBootStateUpdate20(m *Modeenv) (*bootStateUpdate20, error) {
+	u20 := &bootStateUpdate20{}
+	// copy the modeenv for the write object
+	u20.modeenv = m
+	var err error
+	u20.writeModeenv, err = m.Copy()
+	if err != nil {
+		return nil, err
+	}
+	return u20, nil
+}
+
+// commit will write out boot state persistently to disk.
+func (u20 *bootStateUpdate20) commit() error {
+	// The actual actions taken here will depend on what things were called
+	// before commit(), either setNextBoot for a single type of kernel snap, or
+	// markSuccessful for kernel and/or base snaps.
+	// It is expected that the caller code is carefully analyzed to avoid
+	// critical points where a hard system reset during that critical point
+	// would brick a device or otherwise severely fail an update.
+	// There are three things that callers can do before calling commit(),
+	// 1. modify writeModeenv to specify new values for things that will be
+	//    written to disk in the modeenv.
+	// 2. Add tasks to run before writing the modeenv.
+	// 3. Add tasks to run after writing the modeenv.
+
+	// first handle any pre-modeenv writing tasks
+	for _, t := range u20.preModeenvTasks {
+		if err := t(); err != nil {
+			return err
+		}
+	}
+
+	// next write the modeenv if it changed
+	if !u20.writeModeenv.deepEqual(u20.modeenv) {
+		if err := u20.writeModeenv.Write(); err != nil {
+			return err
+		}
+	}
+
+	// finally handle any post-modeenv writing tasks
+	for _, t := range u20.postModeenvTasks {
+		if err := t(); err != nil {
+			return err
+		}
+	}
+
+	// TODO:UC20: reseal against the TPM here with the updated modeenv written
```

Comment:
this TODO is wrongly placed I think it should before the postModenvTasks loop

Reply:
mmm true I just kinda threw the TODO in this function without thought to where the actual call to reseal is, but I can move it around easily enough